providers/printers.rb

```diff
@@ -9,26 +9,33 @@
 # http://www.osor.eu/eupl
 #
 
-def install_printer(prt_name, prt_id, prt_model, prt_uri, prt_policy)
-	print "installing printer #{prt_name}... "
+def create_ppd(prt_name, prt_model, prt_id)
     # foomatic needs all compatible drivers for one printer in order to create the PPD file.
     # But sometimes, there's only one driver available
-	printer_drv = `foomatic-ppdfile -P '#{prt_model}'|grep "'#{prt_id}'"`.scan(/CompatibleDrivers='(\S+)\s.*'/)
-    printer_drv.empty?
-	    printer_drv = `foomatic-ppdfile -P '#{prt_model}'|grep "'#{prt_id}'"`.scan(/Driver='(\S+)*'/)
-	if printer_drv.length >= 0
-		ppd_file=`foomatic-ppdfile -p #{prt_id} -d #{printer_drv[0][0]}`
-		ppd_f = open("/usr/share/cups/model/#{prt_name}.ppd", "w")
-  		ppd_f.write(ppd_file)
-		ppd_f.close
-	else
-		puts "\nthere's no PPD for #{prt_id} in foomatic"
-	end
-	lpadm_comm = Mixlib::ShellOut.new("/usr/sbin/lpadmin  -p #{prt_name} -E -m #{prt_name}.ppd -v #{prt_uri}")
-	lpopt_comm = Mixlib::ShellOut.new("/usr/bin/lpoptions -p #{prt_name} -o printer-op-policy=#{prt_policy} -o auth-info-required=none -o managed-by-GCC=true")
+    Chef::Log.info("Creating PPD file for #{prt_name}... ")
+
+	temp = `foomatic-ppdfile -P '#{prt_model}'|grep "Id='#{prt_id}'"`.scan(/CompatibleDrivers='(.*)'/)
+
+    begin
+        printer_drv = temp[0][0]
+    rescue
+	    temp = `foomatic-ppdfile -P '#{prt_model}'|grep "Id='#{prt_id}'"`.scan(/Driver='(\S+)*'/)
+        printer_drv = temp[0][0]
+    end
+
+    ppd_file=`foomatic-ppdfile -p #{prt_id} -d #{printer_drv}`
+	ppd_f = open("/usr/share/cups/model/#{prt_name}.ppd", "w")
+  	ppd_f.write(ppd_file)
+	ppd_f.close
+	puts "done."
+end
+
+def install_or_update_printer(prt_name, prt_uri, prt_policy)
+    Chef::Log.info("installing or updating printer #{prt_name}... ")
+	lpadm_comm = Mixlib::ShellOut.new("/usr/sbin/lpadmin  -p #{prt_name} -E -m #{prt_name}.ppd -v #{prt_uri} -o printer-op-policy=#{prt_policy} -o auth-info-required=none")
```

Comment:
Presently, changes in PPD_URI are out of control. As we set our ad_hoc option "managed-by-GCC", we should use another option like "external_ppd_uri" to keep the PPD_URI used in the last configuration of the printer. This option should  be saved to the node JSON by OHAI plugin printers.rb (this coincidence of file names can be quirky).